### PR TITLE
[chore][VERSIONING.md] Add long term support policy

### LINKED ISSUE #10004: Define minimum supported period for 1.0
On the [GA roadmap] we stated as a requirement for 1.0 that

> A minimum support period for 1.0 is documented, similarly to [API and SDK](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/versioning-and-stability.md#api-support) stability guarantees.

We likely want to differentiate between Go API guarantees vs behavior guarantees and have the guarantees be per component

### PR DESCRIPTION
**Pull request by mx-psi**
#### Description

Adds support policy for Collector SIG software artifacts with the following principles:
- Support means that we will, at a minimum, fix critical bugs and security issues. We MAY fix other bugs and/or add features if needed (e.g. because a feature eases transition to a v2), but this is not required
- We support artifacts for end-users for longer
- Artifacts may contain sub-components that are not stable. The important bit is that users can easily tell when something is unstable.

#### Link to tracking issue
Fixes #10004

### CODE REVIEW

**Comment by mx-psi** [comment]
cc @open-telemetry/collector-approvers @open-telemetry/collector-contrib-approvers

File: VERSIONING.md
Comment (evan-bradley) [inline VERSIONING.md:32]: Just to be clear, this is elaborating on the previous point and describes a stable version of a component being removed from a distribution with a beta version?

File: VERSIONING.md
Comment (evan-bradley) [inline VERSIONING.md:21]: Do we have the scope of the core framework defined anywhere? Is this just describing the core framework packages used by the components within the distribution?

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md]: I merged it into a single point. The intent is to say that we can remove components from v1 distros, provided we have deprecated the components before

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md]: We have not previously defined it and I am not sure how to best write it. By core framework I mean anything that is not a component that you can plug (so e.g. configuration resolution, service behavior, CLI flags...). Put it another way, the common parts of all OCB builds

File: VERSIONING.md
Comment (TylerHelmuth) [inline VERSIONING.md:26]: How will we handle supported Go versions after 1.0? When 1.24 comes out do we drop support for 1.22 in the 1.0 modules?

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md]: I reworded this to

> The Collector MUST be configurable in a way that does not use unstable components or features.

My point is that we can have, say, alpha components, but one should be able to not use them. This is already the case today: you can not add them to the pipeline, or you can disable a feature gate that you don't want. It was not meant to say anything about what's in the binary

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md]: Called out on 6f3dda2

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md:171]: I feel the other way :smile: at least for modules intended for component developers. We can't have an ecosystem of components outside of contrib if we don't guarantee that the APIs used to build components are stable and available for a long time. Originally I was going to write 2-3 years for component developers, I went with this to start with something a bit more conservative

For the ones meant for library users, I guess 3 months may be fine now because every major user in that audience (vendors and Jaeger) is involved with the Collector

File: VERSIONING.md
Comment (jpkrohling) [inline VERSIONING.md:166]: I feel like this needs a "after v1" remark.

Comment (mx-psi) [reply VERSIONING.md]: Clarified, thanks!

**Review by jpkrohling — APPROVED** [review]

File: VERSIONING.md
Comment (TylerHelmuth) [reply VERSIONING.md:171]: Well I definitely don't want to support the APIs for 3 years, so I can live with 1.

**Review by TylerHelmuth — APPROVED** [review]

**Review by evan-bradley — APPROVED** [review]

File: VERSIONING.md
Comment (codeboten) [inline VERSIONING.md:11]:
```suggestion
End-user focused software artifacts are those that are intended for [end-users](CONTRIBUTING.md#end-users) of the OpenTelemetry Collector. These artifacts include 
```

File: VERSIONING.md
Comment (codeboten) [inline VERSIONING.md:13]: Should this clarify that what is meant by components is the go module that encompass these components?

File: VERSIONING.md
Comment (codeboten) [inline VERSIONING.md:23]: I'm not super clear on what this sentence means, are we talking about users having the ability to build a collector with only stable components?

File: VERSIONING.md
Comment (codeboten) [inline VERSIONING.md:30]: It would be good to have a section in this doc specifically defining what "supported" means

Comment (codeboten) [reply VERSIONING.md:30]: When does support for a binary end?

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md]: Clarified in 78a6621

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md:23]: You should be able to not use any components that are unstable (by not including them in your pipelines) and you should be able to disable feature gates that are unstable. Anything that you can't disable/opt out/not use should be stable. I am not sure how to phrase it better, I appreciate any suggestions

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md:30]: > It would be good to have a section in this doc specifically defining what "supported" means

The intention was that this sentence above:

> [critical bugs](docs/release.md#bugfix-release-criteria) and security vulnerabilities MUST be addressed

was the definition of supported (or, at least, a minimum criteria for what supported means). We may do more than this, but that's the minimum.

> When does support for a binary end?

This was also meant to be already covered in the text by this:

> A binary distribution of the OpenTelemetry Collector MUST be supported for a minimum of **one year** after the release of the next major version of said distribution.

So, basically, we support the distro indefinitely until we release the next major version, after which we support it for one more year.

**Review by atoulme — APPROVED** [review]

File: VERSIONING.md
Comment (tiffany76) [inline VERSIONING.md:166]:
```suggestion
### Long-term support after v1
```

File: VERSIONING.md
Comment (tiffany76) [inline VERSIONING.md:168]:
```suggestion
The OpenTelemetry Collector SIG provides long-term support for stable Go modules. Support for modules depend on the module's [target audiences](CONTRIBUTING.md#target-audiences). The following policies apply to long-term support for any major version starting on v1:
```

**Review by tiffany76 — COMMENTED** [review]
Just a few copy-edit suggestions.

File: VERSIONING.md
Comment (tiffany76) [reply VERSIONING.md:23]: On second thought, “required” doesn’t sound right. So how about

The Collector MUST be configurable so that unstable components or features can be excluded, ensuring that a fully stable configuration is possible.

**Comment by mx-psi** [comment]
> I would add to this document a paragraph that explains that this policy is the current consensus, but it can change in the future, based on project direction and SIG discussions.

Done!

**Review by djaglowski — APPROVED** [review]

**Comment by mx-psi** [comment]
I'll merge this by end of week unless there are further comments, we can iterate on the wording and clarify support more but this feels like it is good enough for a first version given the number of approvals

File: VERSIONING.md
Comment (andrzej-stencel) [inline VERSIONING.md:15]: This mentions Go modules for components. What about all other Go modules published by the project, like e.g. https://pkg.go.dev/go.opentelemetry.io/collector/otelcol or https://pkg.go.dev/go.opentelemetry.io/collector/config/confighttp? Do they get a different treatment than the Go modules for components?

Comment (mx-psi) [reply VERSIONING.md:15]: Yes, they have a smaller support window https://github.com/open-telemetry/opentelemetry-collector/blob/8205489835974f8599e3553e03819ad8a0b6a5db/VERSIONING.md?plain=1#L173 

The rationale for this is that they are equipped to deal with more disruptive changes https://github.com/open-telemetry/opentelemetry-collector/blob/012cf862283a1af411c5b1ace844a6c1d7a82766/CONTRIBUTING.md?plain=1#L45-L46

File: VERSIONING.md
Comment (mx-psi) [reply VERSIONING.md:32]: @codeboten if you think we should file an issue to continue discussing this please do so!

**Pull request merged**
